File: frontend/src/utils/whatsappVentana24h.js
const MS_24H = 24 * 60 * 60 * 1000;

function fechaMensaje(msg) {
  const raw = msg?.creado_en || msg?.created_at || msg?.timestamp || msg?.fecha;
  if (!raw) return null;
  const d = new Date(raw);
  return Number.isNaN(d.getTime()) ? null : d;
}

/** Último mensaje del lead (entrante), excluye sistema y salientes. */
export function getUltimoMensajeEntrante(mensajes) {
  if (!Array.isArray(mensajes) || mensajes.length === 0) return null;

  let ultimo = null;
  let ultimoTs = -Infinity;

  for (const m of mensajes) {
    if (!m || m.direccion !== "entrante") continue;
    const d = fechaMensaje(m);
    if (!d) continue;
    const ts = d.getTime();
    if (ts > ultimoTs) {
      ultimoTs = ts;
      ultimo = m;
    }
  }

  return ultimo;
}

/**
 * @returns {{ abierta: boolean, msTranscurrido: number, msRestante: number, ultimoEntrante: object|null }}
 */
export function calcularVentana24h(mensajes, ahora = Date.now()) {
  const ultimoEntrante = getUltimoMensajeEntrante(mensajes);
  if (!ultimoEntrante) {
    return {
      abierta: false,
      msTranscurrido: MS_24H + 1,
      msRestante: 0,
      ultimoEntrante: null,
    };
  }

  const d = fechaMensaje(ultimoEntrante);
  const ts = d.getTime();
  const msTranscurrido = Math.max(0, ahora - ts);
  const msRestante = Math.max(0, MS_24H - msTranscurrido);

  return {
    abierta: msTranscurrido <= MS_24H,
    msTranscurrido,
    msRestante,
    ultimoEntrante,
  };
}

export function formatTiempoRestante(ms) {
  if (ms <= 0) return "0m";
  const totalMin = Math.ceil(ms / 60000);
  if (totalMin >= 60) {
    const h = Math.floor(totalMin / 60);
    const m = totalMin % 60;
    return m > 0 ? `${h}h ${m}m` : `${h}h`;
  }
  return `${totalMin}m`;
}

export function etiquetaVentanaBadge(ventana) {
  if (ventana.abierta) {
    return `🟢 Ventana abierta · ${formatTiempoRestante(ventana.msRestante)} restantes`;
  }
  return "🔴 Ventana cerrada · pasaron 24h";
}

File: macbot-android/app/src/main/java/com/macbot/app/util/WhatsappVentana24h.kt
package com.macbot.app.util

import android.util.Log
import com.macbot.app.data.api.model.ChatMessage
import java.time.Instant

private const val MS_24H = 24L * 60L * 60L * 1000L
private const val TAG = "WhatsappVentana24h"

data class Ventana24hState(
    val abierta: Boolean,
    val msTranscurrido: Long,
    val msRestante: Long,
    val ultimoEntrante: ChatMessage? = null,
)

/**
 * Misma lógica que [frontend/src/utils/whatsappVentana24h.js] del CRM web.
 * La ventana de atención de WhatsApp se calcula desde el último mensaje **entrante**
 * del cliente (excluye salientes y mensajes de sistema).
 */
fun calcularVentana24h(
    mensajes: List<ChatMessage>,
    ahoraMs: Long = System.currentTimeMillis(),
): Ventana24hState {
    val ultimoEntrante = getUltimoMensajeEntrante(mensajes)
    if (ultimoEntrante == null) {
        return Ventana24hState(
            abierta = false,
            msTranscurrido = MS_24H + 1,
            msRestante = 0,
            ultimoEntrante = null,
        )
    }

    val instant = parseFechaUtcMensaje(ultimoEntrante.creado_en)
    if (instant == null) {
        Log.w(TAG, "No se pudo parsear creado_en del último mensaje entrante")
        // Si no se puede determinar, mantener envío habilitado (comportamiento seguro).
        return Ventana24hState(
            abierta = true,
            msTranscurrido = 0,
            msRestante = MS_24H,
            ultimoEntrante = ultimoEntrante,
        )
    }

    val ts = instant.toEpochMilli()
    val msTranscurrido = (ahoraMs - ts).coerceAtLeast(0)
    val msRestante = (MS_24H - msTranscurrido).coerceAtLeast(0)

    return Ventana24hState(
        abierta = msTranscurrido <= MS_24H,
        msTranscurrido = msTranscurrido,
        msRestante = msRestante,
        ultimoEntrante = ultimoEntrante,
    )
}

/** Último mensaje del lead (entrante), excluye sistema y salientes. */
fun getUltimoMensajeEntrante(mensajes: List<ChatMessage>): ChatMessage? {
    var ultimo: ChatMessage? = null
    var ultimoTs = Long.MIN_VALUE

    for (message in mensajes) {
        if (!message.isEntrante()) continue
        val instant = parseFechaUtcMensaje(message.creado_en) ?: continue
        val ts = instant.toEpochMilli()
        if (ts > ultimoTs) {
            ultimoTs = ts
            ultimo = message
        }
    }

    return ultimo
}

/**
 * PostgREST devuelve timestamptz sin Z; interpretar como UTC (igual que el CRM web).
 */
internal fun parseFechaUtcMensaje(raw: String?): Instant? {
    val value = raw?.trim().orEmpty()
    if (value.isEmpty()) return null

    return try {
        if (value.endsWith("Z", ignoreCase = true) ||
            Regex("[+-]\\d{2}:?\\d{2}$").containsMatchIn(value)
        ) {
            Instant.parse(value)
        } else {
            var iso = value
            if (iso.contains(" ") && !iso.contains("T")) {
                iso = iso.replace(" ", "T")
            }
            Instant.parse("${iso}Z")
        }
    } catch (_: Exception) {
        null
    }
}
